### PR TITLE
Adding 'sequence' to Try and Option

### PR DESCRIPTION
**Pull request by enelson**
Partial fulfillment of #767. Still looking at Either and Match.

### CODE REVIEW

**Comment by danieldietrich** [comment]
Hi Eric, great, thank you! I've already looked into it and will add some comments.

File: src/main/java/javaslang/control/Option.java [hunk 1]
```diff
@@ -46,6 +46,22 @@
...
     /**
+     * Reduces many {@code Option}s into a single {@code Option} by transforming an
+     * {@code Iterable<Option<? extends T>>} into a {@code Option<Seq<T>>}. If any of
+     * the Options are {@link Option.None}, then this returns {@link Option.None}.
+     *
+     * @param values An {@code Iterable} of {@code Option}s.
+     * @param <T> type of the Options.
+     * @return An {@code Option} of a {@link Seq} of results.
+     */
+    static <T> Option<Seq<T>> sequence(Iterable<? extends Option<? extends T>> values) {
```
Comment (danieldietrich) [inline src/main/java/javaslang/control/Option.java]: We provide 'user friendly' messages on NullPointerException, here:

```java
Objects.requireNonNull(values, "values is null");
```

File: src/main/java/javaslang/control/Option.java [hunk 2]
```diff
@@ -46,6 +46,22 @@
     }
 
     /**
+     * Reduces many {@code Option}s into a single {@code Option} by transforming an
+     * {@code Iterable<Option<? extends T>>} into a {@code Option<Seq<T>>}. If any of
+     * the Options are {@link Option.None}, then this returns {@link Option.None}.
+     *
+     * @param values An {@code Iterable} of {@code Option}s.
+     * @param <T> type of the Options.
+     * @return An {@code Option} of a {@link Seq} of results.
```
Comment (danieldietrich) [inline src/main/java/javaslang/control/Option.java]: Accordingly

```java
* @throws NullPointerException if {@code values} is null
```

File: src/main/java/javaslang/control/Option.java [hunk 3]
```diff
@@ -46,6 +46,22 @@
...
+     */
+    static <T> Option<Seq<T>> sequence(Iterable<? extends Option<? extends T>> values) {
+        final Option<Seq<T>> zero = Option.of(List.empty());
+        final BiFunction<Option<Seq<T>>, Option<? extends T>, Option<Seq<T>>> f =
+                (result, option) -> result.flatMap(sequence -> option.map(sequence::append));
+        return Iterator.ofAll(values).foldLeft(zero, f);
```
Comment (danieldietrich) [inline src/main/java/javaslang/control/Option.java]: foldLeft is the most concise solution here. But we cannot shortcut the loop if we visit the first None (see #779).

I would suggest to optimize the internal implementation similar to this:

```java
List<T> list = List.empty();
for (T value : values) {
    if (value.isEmpty) {
        return Option.none();
    }
    list = list.prepend(value.get());
}
return Option.some(list.reverse());
```

I know, from a functional perspective this code looks ugly. But it does perform better on the non-happy path.

File: src/main/java/javaslang/control/Option.java [hunk 4]
```diff
@@ -46,6 +46,22 @@
...
+     * the Options are {@link Option.None}, then this returns {@link Option.None}.
+     *
+     * @param values An {@code Iterable} of {@code Option}s.
+     * @param <T> type of the Options.
+     * @return An {@code Option} of a {@link Seq} of results.
+     */
+    static <T> Option<Seq<T>> sequence(Iterable<? extends Option<? extends T>> values) {
+        final Option<Seq<T>> zero = Option.of(List.empty());
+        final BiFunction<Option<Seq<T>>, Option<? extends T>, Option<Seq<T>>> f =
+                (result, option) -> result.flatMap(sequence -> option.map(sequence::append));
```
Comment (danieldietrich) [inline src/main/java/javaslang/control/Option.java]: Btw - `sequence::append` internally does a `sequence.reverse().prepend(t).reverse()` for every element `t`. Generally use `prepend` instead for every element and finally a `reverse()`.

File: src/main/java/javaslang/control/Try.java [hunk 5]
```diff
@@ -59,6 +62,22 @@
...
     /**
+     * Reduces many {@code Try}s into a single {@code Try} by transforming an
+     * {@code Iterable<Try<? extends T>>} into a {@code Try<Seq<T>>}. If any of
+     * the {@code Try}s are {@link Try.Failure}, then this returns a {@link Try.Failure}.
+     *
+     * @param values An {@link Iterable} of {@code Try}s.
+     * @param <T> type of the Trys.
+     * @return A {@code Try} of a {@link Seq} of results.
+     */
+    static <T> Try<Seq<T>> sequence(Iterable<? extends Try<? extends T>> values) {
```
Comment (danieldietrich) [inline src/main/java/javaslang/control/Try.java]: Also null-check and `@throws` in javadoc

File: src/main/java/javaslang/control/Try.java [hunk 6]
```diff
@@ -59,6 +62,22 @@
...
+        final Try<Seq<T>> zero = Try.of(List::empty);
+        final BiFunction<Try<Seq<T>>, Try<? extends T>, Try<Seq<T>>> f =
+                (result, next) -> result.flatMap(sequence -> next.map(sequence::append));
+        return Iterator.ofAll(values).foldLeft(zero, f);
```
Comment (danieldietrich) [inline src/main/java/javaslang/control/Try.java]: I would rewrite it accordingly to `Option.sequence` as commented above

**Comment by enelson** [comment]
@danieldietrich thanks for all the great feedback! I'll get this all done today.

File: src/main/java/javaslang/control/Option.java (hunk 4 above)
Comment (enelson) [inline src/main/java/javaslang/control/Option.java]: Good to know. So out of curiosity, why do you use `seq::append` in the `sequence()` method for `Future`?

**Comment by danieldietrich** [comment]
Awesome, I'm thankful for you help!

File: src/main/java/javaslang/control/Option.java (hunk 4 above)
Comment (danieldietrich) [inline src/main/java/javaslang/control/Option.java]: It depends on the collection type. The Seq used in Future.sequence is a Stream, which has a lazy append operation taking O(1). It works like a caching iterator. Here we use List as Seq. It has O(1) prepend and O(n) append.

You find a good overview on the bottom of [this page](http://javaslang.com/javadoc/snapshot/javaslang/collection/package-summary.html).

**Comment by enelson** [comment]
@danieldietrich Ok, changes committed for review.

**Comment by danieldietrich** [comment]
Great, thank you!

**Pull request merged**